src/dram_controller.cc

```diff
@@ -552,21 +552,44 @@ bool MEMORY_CONTROLLER::add_wq(const request_type& packet)
   return false;
 }
 
+unsigned long DRAM_ADDRESS_MAPPING::swizzle_bits(champsim::address address, unsigned long segment_size, champsim::data::bits segment_offset, unsigned long field, unsigned long field_bits) const {
+  champsim::address_slice row{get<SLICER_ROW_IDX>(address_slicer), address};
+  unsigned long permute_field = field;
+
+  for(champsim::dynamic_extent subextent{champsim::data::bits{0}, segment_size}; subextent.upper <= row.upper_extent(); subextent = champsim::dynamic_extent{subextent.upper, segment_size}) {
+    permute_field ^= row.slice(subextent).slice(champsim::dynamic_extent{segment_offset,field_bits}).to<unsigned long>();
```

Comment:
I think the double slice is somewhat different than my suggestion. Could you explain how this works? The `address_slice::slice()` family of functions are relative to the domain of the slice. They are not absolute. So, a repeated slice is a subset of a subset of bits. Is that what you intended?

I see you also added some test cases to 751. Are these related?

Reply:
The segment being used from the row address and the bits being xor'd are not identical.

The function travels over the bits of the row address in segments of segment_size.
Bits are taken from [segment_offset + field_bits : segment_offset] of the segment to xor with the field.

The bits chosen for bankgroup and bank are a good example of why this is needed.
Say we have 2 bg bits and 2 bk bits. Zen4 takes the first two bits of the row and xor's it with the first 2 bg bits.
It then takes the next 2 bits from the row and xor's it with the bk bits. It then takes the next two bits and xor's it with the bg bits and so on, back and forth.
So swizzle_bits(address, bk_bits + bg_bits, 0, bankgroup, bg_bits) gives us the bankgroup index
while swizzle_bits(address, bk_bits + bg_bits, bg_bits, bank, bk_bits) gives us the bank index.

The testcases in 751 ensure that these mappings aren't broken. It tests to make sure that for a given row the channel, banks, and bankgroups never alias back into themselves and that each channel, bank, and bankgroup is visited exactly once as the row is traversed.

Reply:
OK, I understand. The segment size is the stride, and the segment offset is the location within that stride. That matches with what's given here, then.